pizza_cutter/coadd_sim_slicer/slicer.py

```diff
@@ -489,40 +504,112 @@ def _build_object_data(
+
+    Returns
+    --------
+    The psf object, currently this will be a PSFEx object
+    """
+    return psfex.PSFEx(fname)
+
+
+def _load_fits_objects(
         *,
         image_path, image_ext,
         bkg_path, bkg_ext,
         weight_path, weight_ext,
         seg_path, seg_ext,
-        bmask_path, bmask_ext,
-        psf):
-    """Read the data.
+        bmask_path, bmask_ext):
+    """
+    Load fits objects for the input paths
+
+    If paths are the same for some objects, the same underlying fits object is
+    used
 
     Returns
     -------
-    img, bkg, seg, wgt, bmask, pex, wcs
+    FITS HDU objects for each type, or None as appropriate
     """
 
-    img = fitsio.read(image_path, ext=image_ext)
-    imh = fitsio.read_header(image_path)
+    im_fits = fitsio.FITS(image_path)
+    im_hdu = im_fits[image_ext]
+
+    imh = im_hdu.read_header()
+
     wcs_dict = {k.lower(): imh[k] for k in imh.keys()}
     wcs = WCS(wcs_dict)
 
-    wgt = fitsio.read(weight_path, ext=weight_ext)
+    # weight map, not optional
+    if weight_path == image_path:
+        wt_fits = im_fits
+    else:
+        wt_fits = fitsio.FITS(weight_path)
 
-    if bkg_path is not None:
-        bkg = fitsio.read(bkg_path, ext=bkg_ext)
+    wt_hdu = wt_fits[weight_ext]
+
+    # bit mask, not optional
+    if bmask_path == image_path:
+        bmask_fits = im_fits
     else:
-        bkg = None
+        bmask_fits = fitsio.FITS(bmask_path)
+
+    bmask_hdu = bmask_fits[bmask_ext]
+
+    # background and seg map, optional
+    bkg_hdu = None
+    if bkg_path is not None:
+        if bkg_path == image_path:
+            bkg_fits = im_fits
+        else:
+            bkg_fits = fitsio.FITS(bkg_path)
+        bkg_hdu = bkg_fits[bkg_ext]
 
+    seg_hdu = None
     if seg_path is not None:
-        seg = fitsio.read(seg_path, ext=seg_ext)
-    else:
-        seg = None
+        if seg_path == image_path:
+            seg_fits = im_fits
+        else:
+            seg_fits = fitsio.FITS(seg_path)
+        seg_hdu = seg_fits[seg_ext]
+
+    return (
+        im_hdu,
+        wt_hdu,
+        bmask_hdu,
+        bkg_hdu,
+        seg_hdu,
+        wcs,
+    )
+
+
+class FakeRandomImage(object):
+    """
+    implements an interface to a random state which can be sliced like an
+    image.  The sigma for normal random numbers is taken from the input weight
+    HDU
+
+    Parameters
+    ----------
+    wgt_hdu: FITS HDU
+        The FITS HDU for the weight map
+    rng: np.random.RandomState
+        A random number generator, to be used to generate random
+        images
+    """
+    def __init__(self, wgt_hdu, rng):
+        self.wgt_hdu = wgt_hdu
+        self.rng = rng
 
-    bmask = fitsio.read(bmask_path, ext=bmask_ext)
+    def __getitem__(self, slices):
 
-    pex = psfex.PSFEx(psf)
+        wt = self.wgt_hdu[slices]
+        sigma = np.sqrt(1.0/wt)
 
-    return img, bkg, seg, wgt, bmask, pex, wcs
+        return sigma*self.rng.normal(size=wt.shape)
```

Comment:
Do we care that if someone asked for two overlapping slices they would get a different noise field out of this object? We may not...

Reply:
good point; I'm not sure a way around this that would not require losing the memory benefits of this approach, except maybe writing the noise slices to a disk cache and reloading as needed

Reply:
We could do the following:

1. divide the image into smallish patches, say 50x50 pixels.
2. save an array of seeds, one for each patch
3. When we get a slice, regenerate the patches need to cover the range of the slice on-the-fly
4. Return the appropriate subset of the data from 3

The only place this matters is in the overlaps between the slices. I don't have a good sense for how important this is.

Reply:
ahh, I had forgotten about overlaps.  Yeah I'm not sure how much it matters

Reply:
I think we should handle this now. We are using this code to test a pretty key part of the pipeline. Seems like we need to do this as right as possible.